mqtt/mqttd/src/app/edgehub.rs

```diff
@@ -0,0 +1,285 @@
+
+use mqtt_bridge::{settings::BridgeSettings, BridgeController};
+use mqtt_broker::{
+    auth::Authorizer,
+    sidecar::{Sidecar, SidecarShutdownHandle},
+    Broker, BrokerBuilder, BrokerHandle, BrokerReady, BrokerSnapshot, FilePersistor,
+    MakeMqttPacketProcessor, Message, Persist, Server, ServerCertificate, SystemEvent,
+    VersionedFileFormat,
+};
+use mqtt_edgehub::{
+    auth::{
+        EdgeHubAuthenticator, EdgeHubAuthorizer, LocalAuthenticator, LocalAuthorizer,
+        PolicyAuthorizer,
+    },
+    command::{
+        AuthorizedIdentitiesCommand, BridgeUpdateCommand, CommandHandler, DisconnectCommand,
+        PolicyUpdateCommand,
+    },
+    connection::MakeEdgeHubPacketProcessor,
+    settings::Settings,
+};
+
+use super::{shutdown, Bootstrap};
+
+const DEVICE_ID_ENV: &str = "IOTEDGE_DEVICEID";
+
+#[derive(Default)]
+pub struct EdgeHubBootstrap {
+    broker_ready: BrokerReady,
+}
+
+#[async_trait]
+impl Bootstrap for EdgeHubBootstrap {
+    type Settings = Settings;
+
+    fn load_config<P: AsRef<Path>>(&self, path: P) -> Result<Self::Settings> {
+        info!("loading settings from a file {}", path.as_ref().display());
+        Ok(Self::Settings::from_file(path)?)
+    }
+
+    type Authorizer = LocalAuthorizer<EdgeHubAuthorizer<PolicyAuthorizer>>;
+
+    async fn make_broker(
+        &self,
+        settings: &Self::Settings,
+    ) -> Result<(Broker<Self::Authorizer>, FilePersistor<VersionedFileFormat>)> {
+        info!("loading state...");
+        let persistence_config = settings.broker().persistence();
+        let state_dir = persistence_config.file_path();
+
+        fs::create_dir_all(state_dir.clone())?;
+        let mut persistor = FilePersistor::new(state_dir, VersionedFileFormat::default());
+        let state = persistor.load().await?;
+        info!("state loaded.");
+
+        let device_id = env::var(DEVICE_ID_ENV).context(DEVICE_ID_ENV)?;
+
+        let authorizer = LocalAuthorizer::new(EdgeHubAuthorizer::new(
+            PolicyAuthorizer::new(device_id, self.broker_ready.handle()),
+            self.broker_ready.handle(),
+        ));
+
+        let broker = BrokerBuilder::default()
+            .with_authorizer(authorizer)
+            .with_state(state.unwrap_or_default())
+            .with_config(settings.broker().clone())
+            .build();
+
+        Ok((broker, persistor))
+    }
+
+    fn snapshot_interval(&self, settings: &Self::Settings) -> std::time::Duration {
+        settings.broker().persistence().time_interval()
+    }
+
+    async fn run(
+        self,
+        config: Self::Settings,
+        broker: Broker<Self::Authorizer>,
+    ) -> Result<BrokerSnapshot> {
+        let mut broker_handle = broker.handle();
+
+        let mut shutdowns = Vec::new();
+        let mut sidecars = Vec::new();
+
+        for sidecar in make_sidecars(&broker_handle, &config)? {
+            shutdowns.push(sidecar.shutdown_handle()?);
+            sidecars.push(tokio::spawn(sidecar.run()));
```

Comment:
do we want to start sidecars *after* the broker?

Reply:
ah, that's the reason we used to spawn server instead of polling it!

Reply:
make broker start first again!